### PR TITLE
Rebind an initializer shared across multiple symbols.

### LINKED ISSUE #53593: VB: Lambda in AsNew clause causes a crash in the compiler
**Version Used**: 

Version 17.0.0 Preview 2.0 [31319.15.d17initial]

**Steps to Reproduce**:

```
Public Class C

    Dim a, b As New C((Function(n) n + 1)(1))

    Sub New(a As Integer) : End Sub
End Class
```

https://sharplab.io/#v2:DYLgbgRgPgkgtgBwPYCcAuBnABAZQJ4ZoCmcAsAFAAKArhMAJYDGWAwsAIYbYsUVb9YAIvThZ2AGiwQsAQWwA5IgHdWAClUAxagDtGaekm2rtASizasAaiwBGE6rsm+A5/xy0sipavazsMbWIAcyIUMxAsAFFtABNcWgpouLZODCA===

```
System.ArgumentException: An item with the same key has already been added. Key: Private Function (n As Object) As Object
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at Microsoft.CodeAnalysis.VisualBasic.LambdaRewriter.Analysis.VisitLambda(BoundLambda node, Boolean convertToExpressionTree)
   at Microsoft.CodeAnalysis.VisualBasic.LambdaRewriter.Analysis.VisitConversion(BoundConversion conversion)
   at Microsoft.CodeAnalysis.VisualBasic.BoundConversion.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitCall(BoundCall node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundCall.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitCall(BoundCall node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundCall.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitObjectCreationExpression(BoundObjectCreationExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundObjectCreationExpression.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitAssignmentOperator(BoundAssignmentOperator node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundAssignmentOperator.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitExpressionStatement(BoundExpressionStatement node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundExpressionStatement.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitSequencePoint(BoundSequencePoint node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundSequencePoint.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitStatementList(BoundStatementList node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundStatementList.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalker.VisitBlock(BoundBlock node)
   at Microsoft.CodeAnalysis.VisualBasic.LambdaRewriter.Analysis.VisitBlock(BoundBlock node)
   at Microsoft.CodeAnalysis.VisualBasic.BoundBlock.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.VisualBasic.BoundTreeWalkerWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.VisualBasic.LambdaRewriter.Analysis.Analyze(BoundNode node)
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Fixes #53593.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_FieldOrPropertyInitializer.vb [hunk 1]
```diff
@@ -54,23 +54,50 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
...
+                    Else
+                        ' Rebind and discard diagnostics
+                        initialValueToRewrite = node.BinderOpt.BindPropertyInitializerExpression(syntax, DirectCast(symbol, PropertySymbol), BindingDiagnosticBag.Discarded)
+                    End If
+                End If
+
+                Dim objectInitializer As BoundObjectInitializerExpression = Nothing
+                Dim createTemporary = True
+
+                If initialValueToRewrite.Kind = Global.Microsoft.CodeAnalysis.VisualBasic.BoundKind.ObjectCreationExpression OrElse initialValueToRewrite.Kind = Global.Microsoft.CodeAnalysis.VisualBasic.BoundKind.NewT Then
```
Comment (chsienki) [inline src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_FieldOrPropertyInitializer.vb:92]: > Global.Microsoft.CodeAnalysis.VisualBasic. [](start = 47, length = 43)

nit: do these need to be fully qualified?

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/LambdaTests.vb [hunk 2]
```diff
@@ -2164,5 +2164,223 @@ In getter
...
+        <WorkItem(53593, "https://github.com/dotnet/roslyn/issues/53593")>
+        <Fact()>
+        Public Sub Issue53593_1()
+
+            Dim compilationDef =
+<compilation>
+    <file name="a.vb">
+Public Class C0
+
+    Public a, b As New C1((Function(n) n + 1)(1))
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/LambdaTests.vb:2176]: >Public a, b As New C1 [](http://example.com/codeflow?start=4&length=21)

Are we testing property initializers as well?

File: src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_FieldOrPropertyInitializer.vb (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_FieldOrPropertyInitializer.vb:92]: > nit: do these need to be fully qualified?

No. Not sure why IDE does this. I'll revert this

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/LambdaTests.vb (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/LambdaTests.vb:2176]: > Are we testing property initializers as well?

Issue53593_6, WithEvents fields are actually properties

**Review by cston — APPROVED** [review]

**Review by chsienki — APPROVED** [review]

**Pull request merged**